cmd/uplink/cmd/cp_test.go

```diff
@@ -0,0 +1,125 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package cmd_test
+
+import (
+	"crypto/rand"
+	"log"
+	"os"
+	"strconv"
+	"testing"
+
+	"storj.io/storj/internal/memory"
+	"storj.io/storj/internal/s3client"
+)
+
+const (
+	bucket                = "testbucket"
+	uplinkEncryptionKey   = "supersecretkey"
+	storjSimSatelliteAddr = "127.0.0.1:10000"
+)
+
+func setup() s3client.Client {
+	var conf s3client.Config
+	conf.Satellite = storjSimSatelliteAddr
+	conf.EncryptionKey = uplinkEncryptionKey
+	conf.APIKey = os.Getenv("storjSimApiKey")
+
+	client, err := s3client.NewUplink(conf)
+	if err != nil {
+		log.Fatalf("failed to create NewUplink: %+v\n", err)
+	}
+	err = client.MakeBucket(bucket, "")
+	if err != nil {
+		log.Fatalf("failed to create bucket %q: %+v\n", bucket, err)
+	}
+	return client
+}
+
+func teardown(client s3client.Client) {
+	for _, bm := range benchmarkCases {
+		deleteUploadedFiles(bm.name)
+	}
+
+	err := client.RemoveBucket(bucket)
+	if err != nil {
+		log.Fatalf("failed to remove bucket %q: %+v\n", bucket, err)
+	}
+}
+
+func deleteUploadedFiles(name string) {
+	for filePath := range uploadedFiles[name] {
+		err := client.Delete(bucket, filePath)
+		if err != nil {
+			log.Printf("failed to delete file %q: %+v\n", filePath, err)
+		}
+	}
+}
+
+var client = setup()
+
+var benchmarkCases = []struct {
+	name     string
+	filesize memory.Size
+}{
+	{"100B", 100 * memory.B},
+	{"1MB", 1 * memory.MiB},
+	{"10MB", 10 * memory.MiB},
+	{"100MB", 100 * memory.MiB},
+	{"1G", 1 * memory.GiB},
+}
+
+var uploadedFiles = map[string]map[string]struct{}{}
+
+func BenchmarkUpload(b *testing.B) {
+	for _, bm := range benchmarkCases {
+		b.Run(bm.name, func(b *testing.B) {
+			file := createRandomBytes(bm.filesize)
+			b.SetBytes(1)
+			b.ResetTimer()
+			for i := 0; i < b.N; i++ {
+				filePath := "folder/data" + strconv.Itoa(i) + "_" + bm.name
+				err := client.Upload(bucket, filePath, file)
+				if err != nil {
+					log.Fatalf("failed to upload file %q: %+v\n", filePath, err)
+				}
+				if uploadedFiles[bm.name] == nil {
+					uploadedFiles[bm.name] = make(map[string]struct{})
+				}
+				uploadedFiles[bm.name][filePath] = struct{}{}
+			}
+		})
+	}
+}
+
+func BenchmarkDownload(b *testing.B) {
+	for _, bm := range benchmarkCases {
+		b.Run(bm.name, func(b *testing.B) {
+			var file string
+			for filePath := range uploadedFiles[bm.name] {
```

Comment:
Good point that we should not write tests that depend on things outside of the test.

The constraints Im trying to account for with the `uploadedFiles` are:
1) the names of each file uploaded need to be unique (because uploading a file that already exists executes different code). I chose to you the value of `b.N` and add that value to the name of the file to create the uniqueness.
2) we need a way to know the name of the files that get uploaded so that we can cleanup at the end and remove them all. 

However regardless of those constraints, you are right, I have built in some dependencies between the tests.  Let me try to remove that.

Reply:
Ok so there is good news and there is bad news.  Good news, I removed the dependency between the upload and download benchmark tests so that they can be run separately now.

The bad news is that caused the tests to take longer because now wee have to create test data for the download benchmark to have something to download. This takes about 30s to upload the test data.

If we dont want this increased time, then another option could be to combine the upload and download benchmarks into one test called `BenchmarkUpDown` or something. Thoughts?

Reply:
Hmm. I assume we're going to run this in an automated way; are we hooking it into Jenkins? I wouldn't really want to add something that time-consuming to our commit verification. But if we can put this somewhere else, I think it's worth the extra time to be able to have separate up and down scores. You _could_ still collect separate up and down scores in a `BenchmarkUpDown`, but I don't think you'd be able to report them in the 'official' Go-benchmark way, so maybe that would make other things hard outside of this. I'm not sure.

I guess all that I feel really strongly about is that we need to collect separate timings for upload and download.

Reply:
I think the goal is to hook this into jenkins **not** related to commit verification. But rather, run benchmarks on a schedule on master branch and store the results somewhere so we can observe over time.  So considering this I think its ok if the tests take a little longer and better to have the upload and download separate.